fendermint/vm/topdown/proof-service/src/assembler.rs

```diff
@@ -0,0 +1,293 @@
+    /// rather than storing it as a field.
+    fn create_client(&self) -> LotusClient {
+        LotusClient::new(self.rpc_url.clone(), None)
+    }
+
+    /// Generate proof bundle for a certificate
+    ///
+    /// Fetches tipsets and generates storage and event proofs.
+    ///
+    /// # Arguments
+    /// * `certificate` - Cryptographically validated F3 certificate
+    ///
+    /// # Returns
+    /// Typed unified proof bundle (storage + event proofs + witness blocks)
+    pub async fn generate_proof_bundle(
+        &self,
+        certificate: &FinalityCertificate,
+    ) -> Result<UnifiedProofBundle> {
+        let generation_start = Instant::now();
+        let instance_id = certificate.gpbft_instance;
+
+        // Get the highest (most recent) epoch from the certificate
+        // F3 certificates contain finalized epochs. On testnets like Calibration,
+        // F3 may lag significantly behind the current chain head (sometimes days).
+        // This can cause issues with RPC lookback limits.
+        let highest_epoch = certificate
+            .ec_chain
+            .suffix()
+            .last()  // Get the most recent epoch in the suffix
+            .or_else(|| certificate.ec_chain.base())  // Fallback to base if suffix is empty
+            .map(|ts| ts.epoch)
+            .context("Certificate has no epochs in suffix or base")?;
+
+        tracing::debug!(
+            instance_id,
+            highest_epoch,
+            "Generating proof bundle - fetching tipsets"
+        );
+
+        // Fetch tipsets from Lotus using proofs library client
+        // We need both parent and child tipsets to generate storage/event proofs:
+        // - Parent tipset (at highest_epoch): Contains the state root we're proving against
+        // - Child tipset (at highest_epoch + 1): Needed to prove state transitions and events
+        //   that occurred when moving from parent to child
+        //
+        // The F3 certificate contains only the tipset CID and epoch, not the full tipset data.
+        // We fetch the actual tipsets here to extract block headers, state roots, and receipts.
+        let client = self.create_client();
+
+        let parent_tipset = client
+            .request(
+                "Filecoin.ChainGetTipSetByHeight",
+                serde_json::json!([highest_epoch, null]),
+            )
+            .await
+            .with_context(|| {
+                format!(
+                    "Failed to fetch parent tipset at epoch {} - RPC may not serve old tipsets (check lookback limit)",
+                    highest_epoch
+                )
+            })?;
+
+        // Child tipset is needed for proof generation - it contains the receipts and
+        // state transitions from the parent tipset
+        let child_tipset = client
+            .request(
+                "Filecoin.ChainGetTipSetByHeight",
+                serde_json::json!([highest_epoch + 1, null]),
+            )
```

Comment:
IIUC, `highest_epoch + 1` is beyond the certified EC chain. How can we prove here anything about the content of `child_tipset`?

Reply:
The child tipset is used to extract receipts and events that occurred during the state transition from parent to child. The actual proof verifies storage/events against the parent tipsets state root (which is certified). The child tipset is just a data source, we are not proving its validity but only what happened during the transition. However, this is a valid concern and we should add comments explaining this clearly.